[1단계 - 자동차 경주 구현] 아벨(신준혁) 미션 제출합니다.

tjdtls690 opened the pull request:
안녕하세요! 백엔드 5기 아벨입니다. 리뷰 잘 부탁드립니다. 감사합니다!

swonny committed:
test: 문자열 분리 테스트

swonny committed:
test: 문자열 삭제 테스트

swonny committed:
test: 특정 인덱스 문자 추출 테스트

swonny committed:
test: 스트링 인덱스 범위 오류 테스트

swonny committed:
test: DisplayName 어노테이션 추가

swonny committed:
test: HashSet 사이즈 테스트

swonny committed:
test: 매개변수 테스트

swonny committed:
test: 매개변수 테스트

swonny committed:
docs: 구현 기능 목록 작성

swonny committed:
feat: 자동차 이름 입력

- 자동차 이름을 입력 받는다.
- 자동차를 생성한다.

swonny committed:
docs: 기능 목록 구현에서 자동차 이름 입력받기 체크

swonny committed:
feat: 시도 횟수 입력 기능 구현

swonny committed:
docs: 구현 기능 목록 체크

swonny committed:
feat: 자동차 이동 기능 구현

swonny committed:
refactor: 자동차 이동 로직 리팩토링

swonny committed:
docs: 기능 구현 목록에서 자동차 이동 기능 체크

swonny committed:
feat: 자동차 위치 출력 기능 구현

swonny committed:
docs: 기능 구현 목록에서 자동차 위치 출력 기능 체크

swonny committed:
fix: 포지션 시작 지점 변경, 출력 포맷 맞추기

swonny committed:
docs: 구현 기능 목록 체크

swonny committed:
feat: 우승자 판별 구현

swonny committed:
docs: 구현 기능 목록 체크 및 수정

swonny committed:
feat: 자동차 이름 5글자 제한 예외 처리 기능 구현

swonny committed:
docs: 자동차 이름 5글자 제한 기능 구현 체크

swonny committed:
feat: 이동횟수가 숫자가 아닌 경우 예외 처리

swonny committed:
docs: 이동 횟수 예외 처리 구현 체크

swonny committed:
refactor: 자동차 이름 예외처리 기능 메서드 분리

swonny committed:
docs: 이동 횟수 입력 예외처리 추가

swonny committed:
feat: 이동횟수 입력 0 이하인 경우 예외 처리

swonny committed:
docs: 이동 횟수 입력 받기 예외 처리 구현 체크

swonny committed:
docs: 기능 구현 목록에 자동차 이름 예외 케이스 추가

swonny committed:
feat: 자동차 이름 입력 시, 쉼표 2개 이상 연속 입력 예외 처리

swonny committed:
docs: 기능 구현 목록에서 자동차 이름 입력 예외 처리 체크

swonny committed:
feat: 자동차 생성자 리팩토링

- 자동차 게임 우승자 테스트를 위한 생성자 추가

swonny committed:
test: 자동차 게임 우승자 구하기 테스트

swonny committed:
docs: 자동차 이름 입력 예외 케이스 추가

swonny committed:
feat: 자동차 한 대 이하로 입력한 경우 예외 처리

- 게임을 진행하기 위해 자동차가 두 대 이상 필요하기 때문에 예외처리를 진행한다.

swonny committed:
docs: 자동차 이름 입력 예외 기능 구현 체크

swonny committed:
test: CarRepository의 findAll 메서드 테스트 추가

- findAll() : 레포지토리에 저장된 모든 자동차 반환하는 메서드

swonny committed:
test: 우승자 판별을 위한 max position 구하기 테스트

swonny committed:
fix: 테스트 시, 레포지토리의 클래스 필드에 자동차가 누적되어 발생하는 오류 해결

swonny committed:
test: 자동차 생성시 이름 길이 예외 처리 테스트

swonny committed:
test: 자동차 개수 예외 테스트

swonny committed:
refactor: 메소드 분리, 불필요한 import 구문 제거, 상수화

Livenow14 commented on docs/README.md:
```diff
@@ -0,0 +1,35 @@
+# 자동차경주 게임
```
요구사항 잘 정리해 주셨네요 👍👍

Livenow14 commented on src/main/java/constant/Constants.java:
```diff
@@ -0,0 +1,16 @@
+    public static final String NAME_JOINING_DELIMITER = ", ";
+    public static final String POSITION_PRINT_FORMAT = "-";
+    public static final String EXCEPTION_STARTER = "[ERROR] ";
+}
```
validation에 사용되는 인스턴스를 제외하고, 전체 애플리케이션에서 1회씩만 사용되고 있어요. 

의미있는 책임 분리였을지 고민해봐도 좋을 것 같아요.

누군가는 이 값을 알기위해 해당 클래스를 몇번이고 반복해서 접근해야할 것 같아서요 :)

tjdtls690 replied:
정말로 지금 다시 보니, 출력 문자열들이 한 번만 사용되기도 하고, 상수 개수가 감당 안될 정도로 많은 것도 아닌데, 굳이 이런식으로 분리하지 말고 연관된 클래스에 두는 것이 더 가독성이 좋았을 수도 있겠다는 생각이 들었습니다.

그리고 여기에 더해서, 한 번만 사용된 문자열들은 상수화 하지 않고, 문자열 그대로 OutputView 클래스의 로직에 나타내는 것으로 다시 리팩토링 했습니다. 이유는 어차피 다른 곳에서 더 쓰이는 것도 아니고, 문자열은 그 내용을 읽어보면 어떤 의미인지 쉽게 파악할 수 있지 않을까 하는 생각 때문이었습니다.

한 가지 질문 드리고 싶은 것은, 리뷰어님은 제가 지금 리팩토링을 진행한 것처럼 한 번만 사용되는 문자열들은 상수화조차도 하지 않고 그냥 그대로 두는 것도 괜찮은지, 아니면 연관된 클래스에 그대로 두면서 상수화는 하는 것이 좋을지 궁금합니다.

Livenow14 replied:
> 한 가지 질문 드리고 싶은 것은, 리뷰어님은 제가 지금 리팩토링을 진행한 것처럼 한 번만 사용되는 문자열들은 상수화조차도 하지 않고 그냥 그대로 두는 것도 괜찮은지, 아니면 연관된 클래스에 그대로 두면서 상수화는 하는 것이 좋을지 궁금합니다


이부분은 개인 취향에 따른 부분이라서요..! 
저는 한 클래스내에 2번 이상 반복된다면, 상수화를 진행해요. 1번만 사용한다면, 로직내에서 그 의미를 충분히 파악할 수 있다고 보거든요 :)

Livenow14 commented on src/main/java/constant/Constants.java:
```diff
@@ -0,0 +1,16 @@
+    public static final String NAME_JOINING_DELIMITER = ", ";
+    public static final String POSITION_PRINT_FORMAT = "-";
+    public static final String EXCEPTION_STARTER = "[ERROR] ";
+}
```
전체 static 인스턴스, 메서드라면, 기본 생성자를 막아도 괜찮을  것 같아요 :)

tjdtls690 replied:
해당 클래스는 리팩토링 하면서 완전히 없어졌지만, 이번에 리팩토링 하면서 생긴 다른 유틸성 클래스들에 private 생성자와 그 생성자 안에서 예외를 던지는 것을 통해, 인스턴스를 생성하지 못하도록 리팩토링 해보았습니다!

Livenow14 commented on src/main/java/controller/GameController.java:
```diff
@@ -0,0 +1,81 @@
+        this.inputView = new InputView(outputView);
+        this.outputView = outputView;
+        this.racingCarGame = new RacingCarGame();
+    }
```
GameController는 scanner와, Racking 게임만 사용할 수 있는걸까요?! 

생성자 내에서 초기화를 한다면, 확장성 부분을 양보하는거라 생각해요

tjdtls690 replied:
생성자에서 초기화 할 때, 확장성 부분을 양보한다는 뜻이 조금 헷갈리는 것 같습니다. 제가 이해한 바로는 inputView, outputView, RacingGame 에 대한 각각의 인터페이스를 전부 만들어서, 그 인터페이스를 인자로 넘겨주는 식으로 구현하는 것으로 확장성을 확보하라는 뜻으로 이해했습니다.

그런데 혹시나 제가 이해한 바가 맞는지 확신이 잘 안서서, 일단 아예 InputView와 OutputView의 메서드를 전부 정적 메서드로 활용하는 것으로 리팩토링 했습니다. 그리고 RacingCarGame은 생성자에서 초기화하는 것으로 그대로 두었습니다.

Livenow14 replied:
> inputView, outputView, RacingGame 에 대한 각각의 인터페이스를 전부 만들어서, 그 인터페이스를 인자로 넘겨주는 식으로 구현하는 것으로 확장성을 확보하라는 뜻으로 이해했습니다.

인터페이스 까지는 아니지만 의도한바는, "게임진행에 필요한 객체들은 생성자의 인자로 넣어주는건 어떨까?" 였습니다. 
controller가 console만 지원한다고 정의했다면, 기존 로직도 나쁘지 않은데요. 

현재 방식에선, inputview가 console이 아닌 다른 방식으로 변경해야한다면, 기존 구현객체에 전체 수정이 이뤄져야하지 않을까요? 
생성자의 인자로 넣는다면, 인자를 변경해주기만 해도 될것 같아서요 !

tjdtls690 replied:
나중에 확장성까지 고려했을 때를 말씀하신 거였군요. 생성자 인자로 전달받는 것으로 리팩토링 해보겠습니다.

Livenow14 commented on src/main/java/controller/GameController.java:
```diff
@@ -0,0 +1,81 @@
+        addCars(inputView.readCarNames());
+        startRacing(readNumberOfMoving());
+        outputView.printWinners(racingCarGame.getWinners());
+    }
```
알맞은 추상화 정도를 선택하신것 같아요 :) 
다만, startRacing과, outputView의 책임이 controller에 있는걸까요? 

아벨이 설정하신 Controller와 RacingCarGame의 역할이 뭔지 알 수 있을까요?!

tjdtls690 replied:
일단 startRacing 은 모든 차를 '입력받은 시도 횟수' 만큼 라운드를 돌려주는 메서드라서, 책임에 맞게 컨트롤러보단 RacingCarGame 에 넣어서 구현하고 싶었습니다. 그런데 여기서 문제가 생겼습니다. 요구사항대로라면 전체 자동차가 한 라운드를 진행할 때마다 차의 포지션들을 출력해줘야 하는데, 그러려면 OutputView 가 필요했습니다.

하지만 컨트롤러는 중개자 역할로서 뷰와 도메인 둘 다 알아도 되지만, 도메인과 뷰는 서로 몰라야 한다고 생각하기 때문에, '도메인 객체인 RacingCarGame' 내에서 OutputView 객체를 통해 자동차들이 한 라운드를 진행할 때마다 자동차들의 포지션을 출력해주는 로직을 짜기가 꺼려졌습니다. 그렇게 되면 **'RacingCarGame (도메인) 이 OutputView (View) 를 알게되는 것이 아닌가??'** 혹은 **'도메인 영역에 View 가 침범해서 도메인 로직에 안 좋은 영향을 끼치지 않을까??'** 하는 생각 때문이었습니다.

그래서 할 수 없이 전체의 자동차들의 라운드를 돌리면서 한 라운드가 끝날 때마다 자동차의 포지션을 출력해주는 로직을 컨트롤러에서 구현하게 되었습니다.

이 부분에 대한 리뷰어님의 의견을 듣고 싶습니다.

Livenow14 replied:
>  전체 자동차가 한 라운드를 진행할 때마다 차의 포지션들을 출력해줘야 하는데, 그러려면 OutputView 가 필요했습니다.

포지션들을 출력하기 위한 방법이 OutputView 뿐일까요? 

도메인에선 자신의 위치 값을 반환하고, 해당 값을 반환받은 controller가 view 필요한 값을 조작해서 반환해준다면, 도메인이 view를 모르게 작성될 수 있을 것 같은데요. 

또한 현재, 작성하신 컨트롤러가 정말 중개자 역할을 하고 있는걸까요? 제가 느끼기엔, main대신 전체 비즈니스 로직과 flow를 담당하는 것으로 보여서요!! 

이부분은 다음 단계 진행하시다 보면 자연스럽게 인사이트를 얻을 수 있을거라 생각해요 :) 너무 급하지 않으셔도 됩니다!!

tjdtls690 replied:
그러면 모든 라운드마다의 각 차들의 포지션을 저장하는 객체를 따로 만들어서, 그 객체를 뷰에 넘겨서 출력하는 형식을 말씀하시는 건가요??

제가 아직 이해가 100% 다 되진 않은 것 같지만, 일단 다음단계 진행하면서 고민해보겠습니다.

Livenow14 commented on src/main/java/controller/GameController.java:
```diff
@@ -0,0 +1,81 @@
+            outputView.printException(illegalArgumentException.getMessage());
+            addCars(inputView.readCarNames());
+        }
+    }
```
addCars 메서드에, makeCar와 try catch까지, 총 3개의 책임이 존재하는것으로 보여요. 
accCars는 단순히, addCars를 할 수는 없는걸까요?

tjdtls690 replied:
일단 예외처리 하는 로직(Controller)과 자동차들을 저장하는 로직(RacingCarGame)을 분리해서 리팩토링을 진행했습니다.

그런데 여기서 한 가지 궁금한 점이 있습니다. 리팩토링 하기 전에 addCars 메서드 안에서 makeCars 메서드를 호출해서 사용 중이었는데, makeCars 는 메서드로 분리되어있는 것이고, addCars메서드가 분리된 메서드를 호출한 것 뿐이니 분리가 된 모습이라고 볼 수 있지 않을까?? 하는 생각이 들었습니다.

물론 try - catch 로 예외처리 하는 부분은 따로 분리해야 한다고 생각하지만, 메서드가 다른 메서드를 호출하는 것이, 호출한 메서드 입장에서 호출된 메서드의 책임까지 전부 떠넘겨받게 되는 것인지 궁금합니다.

Livenow14 replied:
상위 추상화 단계에서 볼까요?

같이 협업을 한다는 가정해서, 제가 구현하지 않은 코드를 볼 때는 비즈니스 로직의 시작과 끝을 상위 추상화 단계에서 주로 확인하게 되는데요. 
ex) NumberGenerate의 "int generate()" 메서드의 상세 구현이 2000자라면, 저는 일단 Number를 int로 만들어 주는구나~ 하고 넘어갑니다. 

 작성하신 코드에서 addCars 메서드는 의미상에서 car를 더한다..! 라고 판단이 들었는데, 상세 구현에서 기대하지 않은 작업들이 많다고 느껴졌어요. 해당 상세 구현을 보지 않고 넘어갔다면, 잠재된 예외를 놓치고 넘어갔을 것 같아서요. 

메서드가 다른 메스드를 호출하는 것이, 과연 정말 addCars를 위한 행동이었지 고민해보면 좋을 것 같아요!

tjdtls690 replied:
아 그렇군요. 결국 '차를 더하는 역할'인 메서드 안에서 '차를 만드는 역할'도 같이 들어있는 것이니, 메서드의 이름으로 예상을 하게 된다면 예상치 못한 로직으로 비춰질 수 있겠군요. 리팩토링 해보겠습니다.

Livenow14 commented on src/main/java/controller/GameController.java:
```diff
@@ -0,0 +1,81 @@
+                .filter(this::validateCarName)
+                .map(Car::new)
+                .collect(Collectors.toList());
+    }
```
filter에서 validation을 처리하고 계시네요..! 
stream.filter의 사용목적은, 스트림내 filtering을 위함인데요.

Stream 클래스의 filter docs를 살펴보면
```java
    /**
     * Returns a stream consisting of the elements of this stream that match
     * the given predicate.
     *
     * <p>This is an <a href="package-summary.html#StreamOps">intermediate
     * operation</a>.
     *
     * @param predicate a <a href="package-summary.html#NonInterference">non-interfering</a>,
     *                  <a href="package-summary.html#Statelessness">stateless</a>
     *                  predicate to apply to each element to determine if it
     *                  should be included
     * @return the new stream
     */
    Stream<T> filter(Predicate<? super T> predicate);

```
Predicate를 받아, Stream내 조건에 일치하는 elements를 반환한다고 되어있어요. 
valdation으로 걸려버리면, stream 자체가 날라가버리는 것 같아서요 :) 
validation 책임을 다른곳으로 옮기는건 어떨까요?

tjdtls690 replied:
그러네요. 확실히 스트림의 filter() 메서드에 맞지않는 역할이면서도 스트림의 효율성을 떨어뜨리는 코드라고 생각 되어서 리팩토링 진행했습니다.

유효성 검증을 하는 로직들을 아예 Validator 클래스로 따로 분리했습니다.

Livenow14 replied:
🤫

Livenow14 commented on src/main/java/controller/GameController.java:
```diff
@@ -0,0 +1,81 @@
+            throw new IllegalArgumentException(COMMA_NOT_ALLOWED_EXCEPTION);
+        }
+        return true;
+    }
```
위에서 말한 부분과 비슷하긴 한데. 
validation이라는 의미는 검증한다! 인데, return의 값을 반환하고 있네요.

- 검증하기 
- 정상적인 값인지 판단해서 참/거짓 반환

의 책임이 들어 있어서요. 


검증과 판단을 분리하는건 어떨까요?

tjdtls690 replied:
맞습니다. 이 부분도 위와 마찬가지로 Validator 클래스로 유효성 검증 책임을 전부 넘기면서 리팩토링을 진행했습니다.

Livenow14 commented on src/main/java/controller/GameController.java:
```diff
@@ -0,0 +1,81 @@
+            outputView.printException(exception.getMessage());
+            return readNumberOfMoving();
+        }
+    }
```
try catch의 책임이 readNumberOfMoving에 있는걸까요?! 

전체 게임을 담당하는 controller에 있을 순 없는걸까요?

tjdtls690 replied:
리팩토링 전의 이 메서드는 현재 컨트롤러에서 getCorrectNumberOfTry() 라는 메서드 명으로 변경하면서 리팩토링을 진행했습니다. getCorrectNumberOfTry() 메서드의 구현부에선, 분리된 readNumberOfTry() 를 통해 입력 값을 받아오고, 메서드의 이름 뜻 처럼, 주 임무인 **'예외 처리를 해서 올바른 입력 값을 반환하는 책임'** 을 갖도록 리팩토링 했습니다.

아니면 혹시 아예 예외 처리하는 메서드를 따로 또 분리해서 

```java
private int getCorrectNumberOfTry() {
        try {
            return readNumberOfMoving();
        } catch (IllegalArgumentException illegalArgumentException) {
            OutputView.printException(illegalArgumentException.getMessage());
            return getCorrectNumberOfTry();
        }
}

private int readNumberOfTry() {
        String numberOfTry = readNumberOfTry();
        NumberOfTryValidator.validate(numberOfTry);
        return Integer.parseInt(numberOfTry);
}
```

이런 식으로 구현하는 것을 얘기하신 건지 궁금합니다.

Livenow14 replied:
예외가 재귀 함수호출용도로 쓰이고 있네요. 
이경우, 해당 함수를 이해하기 위해선, 예외 상황과, 정상적인 상황을 여러번 확인해야할 것 같아요. 
또한, IllegalArgumentException가 아닌 다른 예외가 발생하면 그에따른 플로우는 어떻게 되는걸까요? 


이부분은 "이펙티브자바 69. 예외는 진짜 예외 상황에만 사용하라"를 확인해보시면 좋을 것 같아요!

Livenow14 commented on src/main/java/controller/GameController.java:
```diff
@@ -0,0 +1,81 @@
+        if (numberOfMoving < MINIMUM_NUMBER_OF_MOVING) {
+            throw new IllegalArgumentException(TRIAL_INPUT_EXCEPTION);
+        }
+    }
```
validationTrail이 의미하는바가 모호한 것 같아요. 
검증을 하나만 한다면, method명을 변경하는건 어떨까요? 
 - ex) validateLessThanMininum

tjdtls690 replied:
맞습니다. 메서드 명이 너무 모호해서 메서드의 역할을 파악하기가 좀 힘들다는 생각이 듭니다. 이 부분도 마찬가지로Validator 클래스를 분리하면서 리팩토링을 진행했습니다.

Livenow14 commented on src/main/java/controller/GameController.java:
```diff
@@ -0,0 +1,81 @@
+        for (int round = 0; round < numberOfMoving; round++) {
+            racingCarGame.repeatRounds();
+            outputView.printCarsStatus(CarRepository.findAll());
+        }
```
round 인스턴스가 반복문내에 사용되지 않고 있네요 :) 

```suggestion
    private void startRacing(int numberOfMoving) {
        while (numberOfMoving -- > 0) {
            racingCarGame.repeatRounds();
            outputView.printCarsStatus(CarRepository.findAll());
        }
```
은 어떤가요?

tjdtls690 replied:
사실 가독성 측면에서 위의 for문이 더 낫지 않을까?? 생각했었는데, 리뷰어님 피드백 보고 for문의 round 변수가 아예 사용되지 않고 있다는 부분을 처음 보게 되었네요. 그래서 리뷰어님이 적어주신 while 문으로 리팩토링 진행했습니다.

Livenow14 commented on src/main/java/controller/GameController.java:
```diff
@@ -0,0 +1,81 @@
+    private void startRacing(int numberOfMoving) {
+        for (int round = 0; round < numberOfMoving; round++) {
+            racingCarGame.repeatRounds();
+            outputView.printCarsStatus(CarRepository.findAll());
```
repository에 저장하는 책임이 RacingCarGame에 있을 수  없는걸까요?

tjdtls690 replied:
현실적으로 생각해 봤을 때, 컨트롤러보단 RacingGame 이라는 레이싱 게임 주최가 CarRepository 의 사용 책임을 맡도록 하는 것이 더 옳게 느껴졌습니다.

그래서 리뷰어님 피드백처럼 CarRepository 를 이용해서 자동차 객체를 저장하는 역할을 RacingCarGame 이 맡도록 리팩토링 했습니다.

다만, 위에도 적었지만 RacingCarGame (도메인) 에서 OutputView (View) 를 사용할 수는 없었기에, CarRepository 에서 저장된 자동차들을 꺼내서 경주(라운드를 진행)하는 역할은 맡기지 못했습니다.

Livenow14 replied:
위에 커맨트 달았으니 확인 부탁드려요 :)

Livenow14 commented on src/test/java/racingcar/CarRepositoryTest.java:
```diff
@@ -0,0 +1,62 @@
+
+    @ParameterizedTest
+    @MethodSource("provideCars")
+    void numberOfCarsException(List<Car> cars) {
```
@DisplayName 
혹은 메서드에 한글명으로 작성해보시는건 어떨까요?!

tjdtls690 replied:
해당 어노테이션을 사용하는 것으로 테스트 코드를 전부 리팩토링 해보았습니다! :)

Livenow14 commented on src/test/java/racingcar/CarTest.java:
```diff
@@ -0,0 +1,37 @@
+
+import static org.assertj.core.api.Assertions.*;
+
+class CarTest {
```
도메인 테스트  👍👍

Livenow14 commented on src/test/java/racingcar/RacingCarGameTest.java:
```diff
@@ -0,0 +1,45 @@
+    @MethodSource("provideCars")
+    @DisplayName("메서드를 인수로 전달")
+    void getWinners(List<Car> players, List<String> expectedWinners) {
+        CarRepository.updateCars(players);
+        assertThat(racingCarGame.getWinners()).isEqualTo(expectedWinners);
```
given when then 패턴을 활용하면 가독성이 더 좋아질 것 같아요 :)

tjdtls690 replied:
이 부분에 대해서 궁금한 점이 있습니다.

여기서 given - when - then 으로 나눈다면

```java
given -> 1) 메서드의 인자

when -> 1) CarRepository.updateCars(players);
               2) List<String> winnerNames = racingCarGame.getWinners();

then -> 1) assertThat(winnerNames ).isEqualTo(expectedWinners);
```

이렇게 나눌 것 같은데, 여기서 궁금한 점 2가지가 있습니다.
1. 위의 예시처럼 나누는 것이 맞는 지 궁금합니다.
2. 테스트 코드가 2줄정도만 되는데도 given when then 패턴을 활용하는 것이 더 좋을지도 궁금합니다.

Livenow14 replied:
저는 테스트 코드가 2줄정도만 되도 작성하는 편입니다. 전체 구조화된 테스트가 읽기 편하다 생각하거든요.
(어딘가에선 given-when-then을 사용하고, 어딘가에선 사용하지 않으면, 읽기 힘들어서요.)

다만 이부분은, 개인 취향에 따른 부분이니, 무시하셔도 됩니다 :)

tjdtls690 replied:
그렇군요. 자신이 속한 팀에서 어떤식으로 작성하는지에 따라 달라질 수 있는 부분이겠네요. 다른 방식도 알려주셔서 감사합니다. 연습 해보겠습니다.

Livenow14 commented on src/main/java/racingcar/MoveStrategy.java:
```diff
@@ -0,0 +1,5 @@
+package racingcar;
+
+public interface MoveStrategy {
```
functionalInterface를 잘 설정 해주셨네요 :) 
다만 구현객체가 1개인데, interface화 까지 필요했나 고민해볼 필요가 있을 것 같아요 !

확장성을 고려하다보면 복잡성이 추가되는 경우가 많거든요..!

tjdtls690 replied:
해당 전략 패턴을 사용한 이유는, 자동차의 move() 를 테스트할 때, 랜덤 요소를 전략 패턴으로 빼고, 테스트 코드(CarTest) 에서 보다 더 명확하고 쉽게 테스트를 하기 위해서였습니다.

다만 리뷰어님이 얘기하신 것처럼 확장성 및 복잡성에 관련해서, **'이건 오버 엔지니어링이 아닐까??'** 하는 고민도 들었습니다.

그래도 랜덤의 특성을 명확하게 추출함으로써 각 클래스가 보다 더 명확한 책임을 맡게 되고, 테스트 하기가 수월해지며, 가독성 또한 크게 해치지 않는다고 생각하여 일단 그대로 두었습니다.

Livenow14 replied:
MoveStrategy가 랜덤의 특성을 나타내는걸까요?! 

하나의 전략이라는 것은, 그 전략에 맞게 행동을 취할거라 기대가 되는데요. 

해당 메서드의 return값은 플래그성 변수로 보여서요! 

전략이라는 네이밍 보다, Movement 등으로 변경되도 괜찮지 않을까요?

tjdtls690 replied:
역할(MoveStrategy)이 랜덤의 특성이라기 보단, 이동 전략 중 하나의 전략(구현체)을 랜덤성을 기반으로 한 이동 전략(RandomBasedMoveStrategy)을 만든 것입니다. 나중에 혹시 랜덤을 베이스로 한 이동 전략 외에 '다른 요소를 기반으로 이동한다'는 요구사항이 추가될 시, '해당 요소를 기반으로 한 이동 전략'을 또 추가해 줄 수 있는 형식을 생각했습니다.

물론 나중의 확장성을 생각하지 않고 구현한다면 오버엔지니어링이라 생각 됩니다.

그래도 리뷰어님 의견을 듣고 다시보니, 차라리 숫자를 반환하는 메서드를 최종적인 메서드로 설정하든지, 아니면 리뷰어님이 얘기해주신 것처럼 Movement 로 변경하는 것도 괜찮아 보이네요.

이 부분에 대해서도 더 고민해보겠습니다.

Livenow14 commented on src/main/java/racingcar/MoveStrategy.java:
```diff
@@ -0,0 +1,5 @@
+package racingcar;
+
```
책임이 1개가 있는경우 
java의 @FunctionalInterface 을 활용해보는 것도 좋을 것 같아요

tjdtls690 replied:
함수형 인터페이스엔 항상 이 어노테이션을 붙이는 습관을 갖도록 하겠습니다 :)

Livenow14 commented on src/main/java/racingcar/Car.java:
```diff
@@ -0,0 +1,45 @@
+        validateName(name);
+        this.name = name;
+        this.position = position;
+    }
```
테스트에서만 사용되는 생성자네요. 
이경우 private으로 두고 테스트를 다르게 설정해보셔도 좋을 것 같습니다 :) 

(아벨의 의도는 아니였겠지만, 프로젝트 내에서 어디서든  해당 생성자를 쓰고싶은 욕심이 올라오네요..! )

tjdtls690 replied:
사실 이 부분에 대해서도 굉장히 고민이 많았는데, 만약 이 생성자가 없다면, RacingCarGameTest 테스트 클래스에서 진행되는 테스트가 굉장히 힘들어 질 것 같았습니다.

테스트를 위해 준비할 자동차 객체들을 하나하나 move 를 호출해서 포지션을 설정해줘야하고, 그로 인해 테스트 코드가 너무 감당할 수 없을 만큼 지저분해질 것으로 예상했습니다.

그래서 일단 페어랑 같이 의논한 뒤 이 생성자는 그대로 두기로 했는데, 이에 대한 리뷰어님의 의견이 궁금합니다!

Livenow14 replied:
깔끔한 테스트가 나오지 않다면, 도메인의 책임을 잘못 설정했음을 나타내는 하나의 지표로 볼 수 있을 것 같습니다. 

public은 애플리케이션에서 정말 중요한 비즈니스 로직을 담고 있는게 좋은데요.
그 이유로는, 애플리케이션내의 다른 객체에게 나의 책임을 알리는 수단이기 때문이에요. 

테스트에서만 쓰이는 메서드가 정말 다른 객체에게 알려줘야할 할 책임일까요? 
다른 객체가 해당 객체를 바라볼 때, 놓치면 안될 중요한 메서드일까요? (그랬다면, 비즈니스 로직내에 나타났어야 할 거 같아요.)

이부분은 천천히 많은 고민을 해보시며, 여러 시도를 해보시는게 좋을 것 같아요 :)

Livenow14 reviewed (changes requested):
Test케이스가 있어, 어떤 로직을 가지고 있는지 파악하기 쉬웠어요 :) 
다만, 정말 중요한 로직인지 아닌지를 고민해보고 테스트를 추가/수정/삭제 해보는 경험을 해보시는게 좋을 것 같아요 !! 
(이렇게 하다보면, 객체의 역할과 책임이 더 잘 드러나게 되더라고요)
1단계 구현하시느라 수고많으셨습니다 :)

swonny committed:
refactor: 상수 클래스 삭제

- 한번만 사용되는 상수는 하드코딩으로 변경:
프로그램이 복잡해지는 것을 방지하기 위해 두 번 이상 사용될 때 상수화 하는 방향으로 진행하려고 합니다.

swonny committed:
refactor: 인풋뷰, 아웃풋뷰의 역할 분리

- 인풋뷰가 아웃풋뷰를 가지고 있어서 로직 파악하는데에 힘들기 때문에, 인풋뷰가 아웃풋뷰에 접근하지 못하도록 변경

swonny committed:
refactor: 자동차 이름에 공백이 포함되어 있는 경우 예외 처리

swonny committed:
refactor: 자동차 이름 예외 처리 클래스 분리 및 각 이름에 공백이 포함될 경우 예외처리

swonny committed:
refactor: 이동할 거리를 나타내는 매직넘버 상수화

swonny committed:
refactor: 최대 포지션과 같은 포지션인 차 이름을 반환하는 메서드 이름 개선

swonny committed:
refactor: 랜덤 값을 구현하는 전략패턴의 구현체를 라운드를 반복하는 메서드 안에서 변수로 따로 분리

- 불필요한 객체 생성을 피하기 위해 변수로 분리합니다.

swonny committed:
refactor: 테스트 코드의 가독성을 위해 @Display 어노테이션 추가

swonny committed:
refactor: 자동차 이름 예외 테스트의 @NullAndEmpty 추가 및 자동차 이름 Null 예외 처리 기능 추가

swonny committed:
refactor: CarFactory 클래스 생성 및 컨트롤러에서 자동차 추가하는 메서드 명 개선

swonny committed:
refactor, test: 시도 횟수 예외 처리 클래스 분리, 빈 문자열 예외 처리 추가, 예외 테스트 추가

swonny committed:
refactor(GameController): startRacing 메소드 내부에 사용되지 않는 인스턴스(round) 제거 및 while문으로 변경

swonny committed:
refactor: 함수형 인터페이스에@FunctionalInterface 어노테이션 추가

swonny committed:
test: 자동차 이름 유효성 검사 기능 테스트 추가

swonny committed:
test: 자동차 이름 중복, 자동차 한 대 이하 입력 유효성 검사 기능 테스트 추가

Livenow14 reviewed (approved):
리팩터링 후, 코드가 한결 깔끔해 지셨네요!! 
기존에 남겼던 코멘트에 추가 코멘트 달았는데, 하나의 참고용으로만 봐주시고, 다음 미션 진행해주시면 될 것 같아요..!
(예외 발생시 재귀적 함수 호출 부분은 꼭 수정 부탁드릴게요) 
1단계 수고 많으셨습니다.
